bundle/definition/schema.go

```diff
@@ -0,0 +1,126 @@
+
+	Properties    map[string]*Schema `json:"properties,omitempty" mapstructure:"properties,omitempty"`
+	PropertyNames *Schema            `json:"propertyNames,omitempty" mapstructure:"propertyNames,omitempty"`
+	ReadOnly      *bool              `json:"readOnly,omitempty" mapstructure:"readOnly,omitempty"`
+	Required      []string           `json:"required,omitempty" mapstructure:"required,omitempty"`
+	Then          *Schema            `json:"then,omitempty" mapstructure:"then,omitempty"`
+	Title         string             `json:"title,omitempty" mapstructure:"title,omitempty"`
+	Type          interface{}        `json:"type,omitempty" mapstructure:"type,omitempty"`
+	UniqueItems   *bool              `json:"uniqueItems,omitempty" mapstructure:"uniqueItems,omitempty"`
+	WriteOnly     *bool              `json:"writeOnly,omitempty" mapstructure:"writeOnly,omitempty"`
+}
+
+// SingleType returns true if a schema has a single type declaration
+func (s *Schema) SingleType() bool {
+	_, ok := s.Type.(string)
+	return ok
+}
+
+// GetType will return the singular type for a given schema and a success boolean. If the
+// schema does not have a single type, it will return the false boolean and an error.
+func (s *Schema) GetType() (string, bool, error) {
+	if !s.SingleType() {
+		return "", false, errors.New("this schema has multiple types")
+	}
+	typeString, ok := s.Type.(string)
+	if !ok {
+		return "", false, errors.New("couldn't get type")
+	}
+	return typeString, ok, nil
+}
+
+// GetType will return the types for a given schema and a success boolean. If the
+// schema has a single type, it will return the false boolean and an error.
+func (s *Schema) GetTypes() ([]string, bool, error) {
+	if s.SingleType() {
+		return nil, false, errors.New("this schema a single type")
+	}
+	data, ok := s.Type.([]interface{})
+	if !ok {
+		return nil, false, errors.New("unknown type values")
+	}
+	typeStrings := []string{}
+	for _, val := range data {
+		typeString, ok := val.(string)
+		if !ok {
+			return nil, false, errors.New("unknown type value")
+		}
+		typeStrings = append(typeStrings, typeString)
+	}
+	return typeStrings, ok, nil
+}
+
+// UnmarshalJSON provides an implementation of a JSON unmarshaler that uses the
+// github.com/qri-io/jsonschema to load and validate a given schema. If it is valid,
+// then the json is unmarshaled.
+func (s *Schema) UnmarshalJSON(data []byte) error {
+
+	// Before we unmarshal into the cnab-go bundle/definition/Schema type, unmarshal into
+	// the library struct so we can handle any validation errors in the schema. If there
+	// are any errors, return those.
+	js := new(jsonschema.RootSchema)
+	err := js.UnmarshalJSON(data)
+	if err != nil {
+		return err
+	}
+	// The schema is valid at this point, so now use an indirect wrapper type to actually
+	// unmarshal into our type.
+	type wrapperType Schema
+	wrapper := struct {
+		*wrapperType
+	}{
+		wrapperType: (*wrapperType)(s),
```

Comment:
Is this trick used to avoid the `json.Unmarshal` function to call this very method `func (s *Schema) UnmarshalJSON`, and get stuck into an infinite recursive call?

Reply:
Indeed it is.